[SE-360] Proposed condition clarification and more examples (#1676)
apple/swift-evolution: proposals/0360-opaque-result-types-with-availability.md

## Proposed solution

To bridge this usability gap I propose to relax same-type restriction for functions with `if #available` conditions if they meet all of the following conditions.

1. There is at least one universally (unconditionally) available `return` in a function. All universally available `return`s are required to produce the same type.

2. `if #available` is a top-level statement in a function and *is not* preceded by any block that can return.

  ```swift
  func test() -> some Shape {
    if #available(macOS 100, *) { ✅
      return Rectangle()
    }

    return self
  }
  ```
  
  The following is incorrect because `if #available` is preceded by a dynamic condition that returns:
  
  ```swift
  func test() -> some Shape {
    guard let x = <opt-value> else {
      return ...
    }
    
    if #available(macOS 100, *) { ❌
      return Rectangle()
    }

    return self
  }
  ```

  `if #available` appears inside of a loop:

  ```swift
  func test() -> some Shape {
    for ... {
      if #available(macOS 100, *) { ❌
        return Rectangle()
      }
    }
    return self
  }
  ```

> 3. `if #available` has to terminate with a `return`, and all `return` statements inside of a particular `if #available` condition have to produce the same type, which, at the same time, is allowed be different from top-level `return`(s) or `return`s nested in other conditions regardless of their kind.

Reviewer: ```suggestion
All `return` statements outside of unconditional availability clauses must return the same type as each other, and this type must be as available as the containing function.

All `return` statements within a given unconditional availability clause must return the same type as each other, and this type must be as available as the `#available` condition of the clause.  This type need not be the same type returned by any `return` statement outside of the clause.

There must be at least one `return` statement in the containing function.  If there are no `return` statements outside of unconditional availability clauses, then at least one of the return types within unconditional availability clauses must be as available as the containing function.

Dynamically, the return type of the containing function is:
- the return type of `return` statements in the first unconditional availability clause whose condition is dynamically satisfied, or if none are satisfied then
- the return type of `return` statements outside of all unconditional availability clauses, or if there are no such statements then
- the return type of `return` statements in the first unconditional availability clause that is as available as the containing function.
```

This is slightly more general than before because it allows code like `if #available { return } else { throw ... }`.  Note that that's allowed today as long as the return type is as available as the function.  Basically, the way I've written it, we make a decision tree as long as we syntactically match the pattern, and then maybe the decision tree can be trivialized because every branch produces the same type.

## Proposed solution

To bridge this usability gap, I propose to relax the same-type restriction for functions with `if #available` conditions as follows:

An *unconditional availability clause* is an `if` or `else if` clause that satisfies the following conditions:

  - The clause is part of an `if` statement at the top level of the containing function.
  - There are no `return` statements in the containing function prior to the `if` statement.
  - The condition of the clause is an `#available` condition.
  - The clause is either the initial `if` clause or an `else if` clause immediately following an unconditional availability clause.
  - The clause contains at least one `return` statement.
  - All paths through the block controlled by the clause terminate by either returning or throwing.

> 1. `if #available` is a top-level statement in a function and *is not* preceded by any block that can return.

Reviewer: I don't think you want this to start a list.  This is just restating part of the rule from above while leading into an example; making it part of a list makes the structure confusing, and it can mislead the reader into thinking that you're defining more rules right now.  It's important to be clear about transitions between rules and examples.  Here I would just say something like: "This example satisfies these rules:"

  ```swift
  func test() -> some Shape {
    if #available(macOS 100, *) { ✅
      return Rectangle()
    }

    return self
  }
  ```
  
  The following is incorrect because `if #available` is preceded by a dynamic condition that returns:
  
  ```swift
  func test() -> some Shape {
    guard let x = <opt-value> else {
      return ...
    }
    
    if #available(macOS 100, *) { ❌
      return Rectangle()
    }

    return self
  }
  ```

  Similarly, the following is incorrect because `if #available` appears inside of a loop:

  ```swift
  func test() -> some Shape {
    for ... {
      if #available(macOS 100, *) { ❌
        return Rectangle()
      }
    }
    return self
  }
  ```

> 2. All `return` statements outside of unconditional availability clauses must return the same type as each other, and this type must be as available as the containing function.

Reviewer: We're transitioning back into rules.  I think that's clear enough just because there are no immediate examples.  This should not be part of a list, however.

    - the return type of `return` statements in the first unconditional availability clause whose condition is dynamically satisfied, or if none are satisfied then
    - the return type of `return` statements outside of all unconditional availability clauses, or if there are no such statements then
    - the return type of `return` statements in the first unconditional availability clause that is as available as the containing function.

  The following `test()` function is well-formed because `if` statement produces the same result in both of its branches and it's statically known that the `if #available` always terminates with a `return`
  
  ```swift
  func test() -> some Shape {
    if #available(macOS 100, *) {
       if cond { ✅
         return Rectangle(...)
       } else {
         return Rectangle(...)
       }
    }
    return self
  }
  ```

  But:

  ```swift
  func test() -> some Shape {
    if #available(macOS 100, *) {
       if cond { ❌
         return Rectangle()
       } else {
         return Square()
       }
    }
    return self
  }
  ```

  is not going to be accepted by the compiler because return types are different: `Rectangle` vs. `Square`.

> 3. `if #available` *cannot* appear in an `else` branch unless all preceding clauses are `if #available` conditions that meet all of the requirements listed in this section.

Reviewer: These two examples belongs in the first set of examples, since they are about that first set of rules.  You don't need this lead-in sentence repeating the rule; just a sentence before each example describing why it satisfies or doesn't satisfy the rule is fine.

## Proposed solution

To bridge this usability gap, I propose to relax the same-type restriction for functions with `if #available` conditions: if an `if #available` condition is always executed, it can return a different type than the type returned by the rest of the function.

> Proposed changes allow to:

Reviewer: ```suggestion
The proposed changes allow functions to:
```

Author: Fixed!

> * Use multiple different `if #available` conditions to return types based on their availability e.g. from the most to least available.

Reviewer: ```suggestion
* use multiple `if #available` conditions to return different types based on their dynamic availability and
```

> * Safely fallback to a **single** universally available (meaning no availability restrictions) type if none of the conditions are met.

Reviewer: ```suggestion
* safely fall back to a return type with no availability restrictions if none of the availability conditions are met.
```

Author: Fixed!

> Note that although it is possible to use multiple availability conditions, mixing conditional availability with dynamic checks (`if`, `guard`, `switch`, `for`) would result in `return`s being considered universally available. 

Reviewer: ```suggestion
Because the return type must be decidable without running the code in the function, mixing availability conditions with other conditions (such as `if`, `guard`, or `switch`) removes this special power and requires `return`s in the `if #available` to return the same type as the rest of the function.
```

Author: Fixed!
